From 27e670ef3c91ca1f6146ef0da19df2e8fd5fb938 Mon Sep 17 00:00:00 2001
From: Jan Kratochvil <jan.kratochvil@redhat.com>
Date: Fri, 11 May 2012 18:06:26 +0000
Subject: [PATCH] gdb/
 	Make auto-load handle multiple components of DEBUG_FILE_DIRECTORY.
 	* auto-load.c (auto_load_objfile_script): Remove check for NULL
 	DEBUG_FILE_DIRECTORY.  Handle multiple components of
 	DEBUG_FILE_DIRECTORY.

---
 gdb/ChangeLog   |    7 +++++++
 gdb/auto-load.c |   32 ++++++++++++++++++++++----------
 2 files changed, 29 insertions(+), 10 deletions(-)

Index: gdb-7.2/gdb/auto-load.c
===================================================================
--- gdb-7.2.orig/gdb/auto-load.c
+++ gdb-7.2/gdb/auto-load.c
@@ -587,17 +587,29 @@ auto_load_objfile_script (struct objfile
   input = fopen (filename, "r");
   debugfile = filename;
 
-  if (!input && debug_file_directory)
+  if (!input)
     {
-      /* Also try the same file in the separate debug info directory.  */
-      debugfile = xmalloc (strlen (filename)
-			   + strlen (debug_file_directory) + 1);
-      strcpy (debugfile, debug_file_directory);
-      /* FILENAME is absolute, so we don't need a "/" here.  */
-      strcat (debugfile, filename);
-
-      make_cleanup (xfree, debugfile);
-      input = fopen (debugfile, "r");
+      char *debugdir;
+      VEC (char_ptr) *debugdir_vec;
+      int ix;
+
+      debugdir_vec = dirnames_to_char_ptr_vec (debug_file_directory);
+      make_cleanup_free_char_ptr_vec (debugdir_vec);
+
+      for (ix = 0; VEC_iterate (char_ptr, debugdir_vec, ix, debugdir); ++ix)
+	{
+	  /* Also try the same file in the separate debug info directory.  */
+	  debugfile = xmalloc (strlen (debugdir) + strlen (filename) + 1);
+	  strcpy (debugfile, debugdir);
+
+	  /* FILENAME is absolute, so we don't need a "/" here.  */
+	  strcat (debugfile, filename);
+
+	  make_cleanup (xfree, debugfile);
+	  input = fopen (debugfile, "r");
+	  if (input != NULL)
+	    break;
+	}
     }
 
   if (!input && gdb_datadir)
